paasta_tools/kubernetes/bin/paasta_secrets_sync.py

```diff
@@ -245,63 +253,89 @@ def sync_secrets(
-                        )
-                    else:
-                        log.info(f"{secret} for {service} in {namespace} up to date")
+
+                update_k8s_secret(
+                    service=service,
+                    secret_name=f"{namespace}-secret-{service}-{sanitise_kubernetes_name(secret)}",
+                    secret_data={
+                        secret: base64.b64encode(
+                            secret_provider.decrypt_secret_raw(secret)
+                        ).decode("utf-8")
+                    },
+                    secret_signature=secret_signature,
+                    kube_client=kube_client,
+                    namespace=namespace,
+                )
+
+    return True
+
+
+def sync_crypto_secrets(
+    kube_client: KubeClient,
+    cluster: str,
+    service: str,
+    secret_provider_name: str,
+    vault_cluster_config: Dict[str, str],
+    soa_dir: str,
+    vault_token_file: str,
+) -> bool:
+    """
+    For each key-name in `crypto_key`,
+    1. Fetch all versions of the key-name from Vault superregion mapped from cluster, e.g. `kubestage` maps to `devc` Vault server.
+    2. Create K8s secret from JSON blob containing all key versions.
+    3. Create signatures as K8s configmap based on JSON blob hash.
+
+    So each instance of a service get the same key, thereby reducing requests to Vault API
+    """
+    config_loader = PaastaServiceConfigLoader(service=service, soa_dir=soa_dir)
+    for instance_config in config_loader.instance_configs(
+        cluster=cluster, instance_type_class=KubernetesDeploymentConfig
+    ):
+        crypto_keys = instance_config.get_crypto_keys_from_config()
+        if not crypto_keys:
+            continue
+        secret_data = {}
+        provider = get_secret_provider(
+            secret_provider_name=secret_provider_name,
+            soa_dir=soa_dir,
+            service_name=service,
+            cluster_names=[cluster],
+            secret_provider_kwargs={
+                "vault_cluster_config": vault_cluster_config,
+                "vault_auth_method": "token",
+                "vault_token_file": vault_token_file,
+            },
+        )
+        for key in crypto_keys:
+            key_versions = list(provider.get_key_versions(key))
+            if not key_versions:
+                log.error(
+                    f"No key versions found for {key} on {instance_config.get_sanitised_deployment_name()}"
+                )
+                continue
+
+            secret_data[get_vault_key_secret_name(key)] = base64.b64encode(
+                json.dumps(key_versions).encode("utf-8")
+            ).decode("utf-8")
+
+        if not secret_data:
+            continue
+
+        # In order to prevent slamming the k8s API, add some artificial delay here
+        time.sleep(0.3)
+        update_k8s_secret(
+            service=service,
+            # `kubernetes.client.V1SecretVolumeSource`'s `secret_name` must match `secret` below
```

Comment:
editing this to include what you mentioned at https://github.com/Yelp/paasta/pull/3536#discussion_r1148101728 would make this comment clearer (imo)

e.g.,
```suggestion
            # the secret name here must match the secret name given in the secret volume config
```
(or even your comment as-is :))